Zoi Kousteni - Dog Lover App

Lunfer opened the pull request:
Here is my NextJS app called "Dog Lover" with all of the requested views included as well as the search input field (noted as a "nice to have"). When you "npm run dev", I hope that you will enjoy my humour as much as I do :D

Lunfer committed:
created next app - initial commit

Lunfer committed:
first take of the header

Lunfer committed:
leftSide box and main grid created

Lunfer committed:
added the dogGenerator box

Lunfer committed:
added the BreedSelector box

Lunfer committed:
almost done, need to fix some react router stuff first

Lunfer committed:
DONE! maybe I should double check it though

Lunfer committed:
double_ckecking part 1 done

Lunfer committed:
fine tuned and with a search field fully functioning

padelis commented on dog_lover/pages/api/hello.ts:
```diff
@@ -0,0 +1,13 @@
+// Next.js API route support: https://nextjs.org/docs/api-routes/introduction
+import type { NextApiRequest, NextApiResponse } from 'next'
```
Is this used somewhere?

Lunfer replied:
whoops nope! forgot to delete that one!

padelis commented on dog_lover/src/breedContext.tsx:
```diff
@@ -0,0 +1,103 @@
+  return useContext(QueryContextUpdate);
+};
+
+export const BreedProvider = ({ children }: { children: any }) => {
```
How did you named this BreedProvider? I mean it's not a Component?

Lunfer replied:
Well, it's my first time with the context API in general, I learned all I needed to know about it through an awesome YouTuber called Web Dev Simplified, so I don't quite understand your question: is it usually impossible to change its name? To be honest, now that I see it, I think I should have named it "DogProvider", it makes much more sense :D

padelis replied:
My question was derived from the file name. I expected to see context related code here, and the rest in a different file like `DogProvider` or `Dogs` under the components or pages directory.

Lunfer replied:
Yes, that would make sense. I am without a doubt new with the Context API so I am not exactly familiar with the best practices around it.

padelis commented on dog_lover/src/breedContext.tsx:
```diff
@@ -0,0 +1,103 @@
+    const keys = Object.keys(allBreeds);
+
+    // resolve images and create DogBreed objects
+    const dogBreeds = await Promise.all(
```
What will happen if one of the promises fail?

Lunfer replied:
You are absolutely right, I thought I could get away with not adding a try/catch! Will do it in my next commit if I am allowed to of course.

padelis commented on dog_lover/components/topBar.tsx:
```diff
@@ -0,0 +1,262 @@
+                    fontWeight: "600",
+                  }}
+                >
+                  You are in the right place.
```
🤣  LOL

padelis reviewed (commented):
Hey @Lunfer!

Thank you for taking the time to do our assignment!
I've added some comments in your PR, please have a look.

Also have you ever tried any kind of testing library?

Lunfer commented:
> Hey @Lunfer!
> 
> Thank you for taking the time to do our assignment! I've added some comments in your PR, please have a look.
> 
> Also have you ever tried any kind of testing library?

Unfortunately, I haven't yet, I would love to though! I've heard of a few testing frameworks with the most popular being Cypress, especially in the "NextJS enthusiasts' circle"!

Lunfer committed:
cleaned up my code

*removed unnecessary file hello.ts
*made boxes DogGenerator and BreedSelector a little more functional

teliaz commented on dog_lover/components/breedImageGenerator.tsx:
```diff
@@ -0,0 +1,87 @@
+        <ImageList
+          gap={1}
+          sx={{
+            width: 500,
```
How come you don't use classes instead of styles?

Lunfer replied:
Nice question! So I understand that my code would look MUCH cleaner if I used classes in a separate CSS file, *but* ever since I came across Material UI, I just loved the fact that I can do anything I want right there inside my components file, with no need of just going on my other file.css and adjusting the styles etc. Since I have my app on my browser compiling/rendering/refreshing at the same moment, for me, it feels like drawing. Call it a personal preference then. To be honest, since I am working on storybook.js at the moment, I'm used to using module.css files inside my components so I thought I could do it differently now that I have the chance!

teliaz commented on dog_lover/components/dogGeneratorBox.tsx:
```diff
@@ -0,0 +1,73 @@
+  useVariableRandomImageContext,
+  useFunctionRandomImageContext,
+} from "../src/breedContext";
+// import "aos/dist/aos.css";
```
It this needed here?

Lunfer replied:
Ah yes, these comments that are left along with the respective package in the package.json, is my *failed* attempt to add it to my Box components to make them more interactive if I may say so, or just animate them. I loved this package when I was designing my own website and wanted to use it here as well, but it was acting up! I think the latest version of it is in desperate need of an update/bug fix! And of course, leaving those comments was actually on purpose because I really wanted to talk about this package hahaha it made my life so much easier when I wanted to animate my website!

teliaz commented on dog_lover/components/topBar.tsx:
```diff
@@ -0,0 +1,262 @@
+                  disableElevation
+                  onClick={twoClicksBreedsList}
+                  sx={{
+                    boxShadow: "none",
+                    background: "rgba(103, 164, 255, 0.7)",
+                    padding: "7px",
+                    border: "1px solid black",
```
Some of the styles here are repeated. Could you have implemented this some other way?

Lunfer replied:
YES definitely with classes (AKA classNames in React). Would look much easier on the eye as well!

teliaz commented on dog_lover/package.json:
```diff
@@ -0,0 +1,31 @@
+    "react-router-dom": "^6.2.1"
+  },
+  "devDependencies": {
+    "@types/aos": "^3.0.4",
```
Are you using this dependency?

teliaz commented on dog_lover/src/breedContext.tsx:
```diff
@@ -0,0 +1,135 @@
+
+  return (
+    <ClickedBreedContext.Provider value={clickedBreed}>
+      <BreedContext.Provider value={breeds}>
+        <HeartContextUpdate.Provider value={handleHeartClicked}>
+          <FetchBreedsContextUpdate.Provider value={fetchBreedsList}>
+            <OpenContext.Provider value={isOpened}>
+              <QueryContext.Provider value={query}>
```
It seems that you are using each provider for a single purpose. If you had to combine them how would you approach this?

Lunfer replied:
So this is my first attempt at using Typescript AND Context API, so when I tried to combine them, which I did at first, I just added all of the variables inside one provider and the functions inside the other, but I was getting errors that I think were mainly because not all of them have the same types. So I think the way to go would be to combine them if they had the same interfaces/types. I am not sure about that one though!

teliaz commented on dog_lover/components/topBar.tsx:
```diff
@@ -0,0 +1,262 @@
+    }
+  };
+  const oneClickBreedsList = () => {
+    const breedSelectorButton = document.getElementById("whoLetTheDogsOut");
```
Nice naming here! LGTM!

Lunfer replied:
Thanks! Also thank you for the "LGTM", I had no idea what it meant but now I do 😂

teliaz reviewed (commented):
Thanks for the effort you put into this test! Can you please answer the following alongside my comments:

- It was creative indeed, but how come you didn't have routes/views? 
- Why did you use nextJS for this project? 

Really creative test!!

Lunfer commented:
> Thanks for the effort you put into this test! Can you please answer the following alongside my comments:
> 
> * It was creative indeed, but how come you didn't have routes/views?
> * Why did you use nextJS for this project?
> 
> Really creative test!!

*Okay, I am in shock right now, I really misunderstood the whole "view" thing and for a moment I thought you used it just a fancy way to describe either pages or components. So that part is completely on me and of course, I apologise about it. *If you would like I could easily swap it with my Box tags.* 

*Also, I used NextJS because I really like it and the way it just contains everything or almost everything I could need when making an app. I am also more familiar with it and, I don't know if that's a true statement or not, but it feels more like a full-stack approach to creating something like I have control over everything I do inside of it. 

I hope that answers your questions! Thank you for your compliment and for your time and effort!
